subiquity/controllers/error.py

```diff
@@ -0,0 +1,177 @@
+# Copyright 2019 Canonical, Ltd.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Affero General Public License as
+# published by the Free Software Foundation, either version 3 of the
+# License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Affero General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+import datetime
+import enum
+import json
+import logging
+import os
+
+import apport
+import apport.crashdb
+import apport.hookutils
+
+import attr
+
+from subiquitycore.controller import BaseController
+
+
+log = logging.getLogger('subiquity.controllers.error')
+
+
+class ErrorReportState(enum.Enum):
+    INCOMPLETE = _("INCOMPLETE")
+    DONE = _("DONE")
+    ERROR = _("ERROR")
+
+
+class ErrorReportKind(enum.Enum):
+    BP_FAIL = _("Block device probe failure")
+    RBP_FAIL = _("Disk probe failure")
+    INSTALL_FAIL = _("Install failure")
+    UI = _("Installer crash")
+    UNKNOWN = _("Unknown error")
+
+
+@attr.s(cmp=False)
+class ErrorReport:
+    controller = attr.ib()
+    base = attr.ib()
+    pr = attr.ib()
+    state = attr.ib()
+    _file = attr.ib()
+
+    meta = attr.ib(default=attr.Factory(dict))
+
+    @classmethod
+    def new(cls, controller, kind):
+        prefix = "installer.{}.{}".format(
+            datetime.datetime.utcnow().isoformat(timespec='seconds'),
+            kind.name.lower())
+        i = 0
+        while 1:
```

Comment:
This should have some sort of upper-bound and timeout.

Reply:
How about an upper bound of 10k? :)